WIP: Esp32 support

xermicus opened the pull request:
Run sled on ESP32 boards.

This is still WIP!

vifino commented on src/os/os_freertos.c:
```diff
@@ -0,0 +1,117 @@
+void oscore_task_pin(oscore_task task, int cpu) {}
+
+void * oscore_task_join(oscore_task task) {
+	// ye ok
```
We probably want to implement some way of joining, but it's not required, only really for clean exits (which aren't really applicable to mcus..)

xermicus replied:
I gave it a shot (but didn't test yet).

vifino replied:
It'll do for now, I think.

vifino commented on Makefiles/esp32.GNUmakefile:
```diff
@@ -0,0 +1,220 @@
+endif
+
+
+# Default configuration starts here. This all uses ?=, so SLEDconfig makes the final decision.
```
Pretty much none of those default settings need to be here, these are supposed to be overrides of specific settings.
These are included after the settings if i recall correctly. 
None of these will have any effects, because it'll simply be set before.
Check `GNUmakefile`.

vifino commented on Makefiles/esp32.GNUmakefile:
```diff
@@ -0,0 +1,220 @@
+# Available: 'unix', possibly '3ds'
+PLATFORM ?= unix
+
+DEFAULT_OUTMOD ?= sdl2
```
You probably don't want this.

vifino commented on Makefiles/esp32.GNUmakefile:
```diff
@@ -0,0 +1,220 @@
+CFLAGS += -DDEFAULT_OUTMOD=\"$(DEFAULT_OUTMOD)\" -DDEFAULT_MODULEDIR=\"$(DEFAULT_MODULEDIR)\"
+
+SOURCES := src/asl.c      src/main.c        src/mod.c
+SOURCES += src/matrix.c   src/random.c      src/timers.c  src/util.c
```
These will all be set in the main GNUmakefile.

xermicus commented:
Cool, thank you much for the thorough review! It's really WIP and your corrections/suggestion are very welcome :-)
I'll later try to add sled to platformio as a library. This would mean you can build it with make without having to worry about deps, and later use it as a library. platformio will keep track of all freertos and other dependencies and also link it correctly. This way it will be very easy for others to use.

vifino commented:
Also, any generic building things, like linking into a library, should be in the main GNUmakefile, esp32.GNUmakefile should contain only esp32 specific overrides.

By the way, you don't have to implement all of these, most of them are suggestions and if the need arises to implement them, all of the above I can do myself. Just lack some testing hw at the moment. :)

xermicus commented on Makefiles/esp32.GNUmakefile:
```diff
@@ -0,0 +1,220 @@
+
+# The following lists are for reference.
+
+GFXMODS_AVAILABLE := gfx_random_static gfx_random_rects gfx_twinkle gfx_gol
```
ack

xermicus commented on Makefiles/esp32.GNUmakefile:
```diff
@@ -0,0 +1,220 @@
+endif
+
+libsled.a: $(OBJECTS) $(ML_OBJECTS)
+	 $(AR) rcs "$@" $(filter-out os_dummy.o os_unix.o,$^)
```
I analyzed the compiled (static) result and noticed that those are somehow compiled into it too. But you probably don't need it and space on those dev boards is sometimes sparse.

vifino replied:
This shouldn't happen. If it compiles other os_* things than the one wanted, something else is going wrong.

vifino replied:
It's probably not a bad idea to move the `libsled.a` rule into the Makefile itself, but without the filter.
Not sure if you can figure out why it happens.

xermicus commented on Makefiles/sledconf.esp32:
```diff
@@ -0,0 +1,59 @@
+CFLAGS += -fdata-sections
+CFLAGS += -fstrict-volatile-bitfields
+# TODO add dependencies to ext:
+CFLAGS += -I/home/cyrill/.platformio/packages/framework-arduinoespressif32/tools/sdk/include/freertos/
```
What about something like ${PLATFORMIO_HOME} that can be set in the sledconf?

vifino replied:
Yes, that's a good idea. Make it default to ${HOME}/.platformio in sledconf.esp32.

xermicus commented on Makefiles/sledconf.esp32:
```diff
@@ -0,0 +1,59 @@
+CFLAGS += -I/home/cyrill/.platformio/packages/framework-arduinoespressif32/tools/sdk/include/esp_ringbuf/
+
+# Use the xtensa toolchain
+CC := xtensa-esp32-elf-cc
```
I installed it via package manager. But probably platformio ships those too (shouldn't matter I think).

vifino replied:
Well, it's probably a good idea to use the same toolchain used to build the other stuff.
Luckily it's not so much of a problem as it might've been with C++ code.
Probably a good idea to default to platformio's toolchain for consistency.

xermicus replied:
ACK. I have the toolchain available also in the ~/.platformio path so I assume it ships with it. We could do something like that?
```suggestion
CC := ${PLATFORMIO_HOME}/packages/toolchain-xtensa32/bin/xtensa-esp32-elf-cc
AR := ${PLATFORMIO_HOME}/packages/toolchain-xtensa32/bin/xtensa-esp32-elf-ar
```

vifino replied:
Yeah, sounds good.

xermicus commented on src/modules/out_esp32.c:
```diff
@@ -0,0 +1,101 @@
+}
+
+int get(int _modno, int x, int y) {
+	return ppos(x, y);
```
lol you're right. your implementation seems fine. thx!

orithena commented:
@vifino Dangit, if I knew this last sunday... I could've given you an ESP32.

vifino commented:
@orithena Heh, I should have an ESP32 somewhere, but right now I lack 5V PSU's and ws2812b matrices that don't suck.

.. Actually, I still have a tshirt I've messed with..

vifino commented:
Since platformio probably has support for the esp8266 - which is also xtensa i believe - and can also run freertos, it might be pretty simple to add support for it as well.

The output module probably has to be replaced, but that's little effort in comparison to figuring out how to get it to compile. See #26 for my failed efforts.

xermicus commented:
Yes just install platformio and then `platformio platform install espressif8266`. Also I see a esp8266-freertos-sdk framework when I do `platformio platform frameworks` so everything you need should be provided. Not sure how to install the esp8266-freertos-sdk framework from the command line though but you can probably just use a dummy project to let the dependency finder install the framework...

vifino commented:
Hey, if you could write some (hopefully detailed) instructions on how to build this for ESP32's, that'd be nice.
Since I'm home now, I got ws2812b matrices and esp32's, so testing this would be cool.
It'd also help me with doing CI.

vifino commented on src/os/os_freertos.c:
```diff
@@ -0,0 +1,121 @@
+	if (core < 0 || core >= oscore_ncpus()) {
+		xTaskCreate(user_main, "sled_main", STACK_DEPTH, pvParameters, TASK_PRIORITY, NULL);
+	} else {
+		xTaskCreatePinnedToCore(user_main, "sled_main", STACK_DEPTH, pvParameters, TASK_PRIORITY, NULL, core);
```
If you could rename the the sled_task function to user_main (or whatever your entry point is) and also rename the user_main function to sled_task, that'd be nice.
They are still confusing.

xermicus commented:
Sure, I'm working on a Dockerfile for our project where we need seld. Ill attach it here when ready.

vifino commented:
Alright, cool.

xermicus commented:
you can build it with pio now:
```Dockerfile
FROM archlinux/base

RUN pacman -Sy
RUN pacman -Syyu --noconfirm
RUN pacman -S --noconfirm git sudo binutils coreutils python2 python2-pip tar file awk fakeroot make 
RUN useradd -m build
RUN echo "build ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers

USER build
WORKDIR /home/build
RUN git clone https://aur.archlinux.org/python-click-5.1.git
RUN git clone https://aur.archlinux.org/platformio.git
RUN git clone --single-branch --branch esp32-support https://github.com/xermicus/sled.git
RUN git clone --single-branch --branch platformio-library https://github.com/xermicus/ESP32-Digital-RGB-LED-Drivers.git

RUN cd python-click-5.1 && makepkg -si --noconfirm
RUN cd platformio && makepkg -si --noconfirm

RUN pio platform install espressif32 --with-package framework-arduinoespressif32
RUN pio lib --global install /home/build/ESP32-Digital-RGB-LED-Drivers/

RUN cd sled && \
    cp Makefiles/sledconf.esp32 sledconf && \
    make libsled.a
```

vifino commented on Makefiles/sledconf.esp32:
```diff
@@ -53,7 +53,5 @@ CFLAGS += -I${PLATFORMIO_HOME}/packages/framework-arduinoespressif32/tools/sdk/i
-AR := xtensa-esp32-elf-ar
-
+# led lib for out mod
+CFLAGS += -I${PLATFORMIO_HOME}/lib/ESP32-Digital-RGB-LED-Drivers/include/
```
You can actually do this a bit better I believe.

You can try adding a `.incs` file with that `-I${PLATFORMIO_HOME}/lib/ESP32-Digital-RGB-LED-Drivers/include/`, it should propagate. Not sure if Make exports those variables, though.

xermicus replied:
Yes that would be nice but I did some tests and it looks like it won't get exported?

vifino replied:
That's weird, it should.

vifino commented:
Nice, I'll add a CI config once I'm home from work.
How do I get to the final binary which I can flash and use?

vifino commented on Makefiles/esp32.GNUmakefile:
```diff
@@ -1,220 +1,2 @@
-	$(CC) $(CPPFLAGS) $(CFLAGS) $(LDFLAGS) -o $@ $^ $(LIBS) `cat $(PLATFORM_LIBS) $(MODULES_STATIC_LIBS) 2>/dev/null || true`
-endif
-
 libsled.a: $(OBJECTS) $(ML_OBJECTS)
```
You can probably put the rule in the main file.

vifino commented on Makefiles/esp32.GNUmakefile:
```diff
@@ -0,0 +1,5 @@
+CC := ${PLATFORMIO_HOME}/packages/toolchain-xtensa32/bin/xtensa-esp32-elf-cc
```
Note that this is the reason the CI is failing.

xermicus replied:
yep, thx!

xermicus commented on Makefiles/sledconf.esp32:
```diff
@@ -0,0 +1,57 @@
+MATRIX_Y := 9
+
+## pixel size
+SDL_SCALE_FACTOR := 1
```
ack

xermicus commented on src/os/os_freertos.c:
```diff
@@ -0,0 +1,121 @@
+
+// -- event
+oscore_event oscore_event_new(void) {
+	return NULL;
```
Yes I think we should try to. I'll look into it today evening.

xermicus commented:
Cool! No worries I'll fix all the minor things that I haven't yet.
As for the firmware build I can look into it after work. Do you prefer a makefile target to build something flashable or completely depend on platformio? Doing it via a platformio project would be very easy but I think both is possible, I'll probably try both. I'll also check how easy it would be to add sled to platformio registry to see whether a separate repository is necessary,

xermicus commented:
Removed the filter in libsled.a target, works fine now.
Was probably due to some stray .o files laying around or something.

xermicus commented:
Switched the outmodule to `out_esp32_digitalledlib.c` and included a check in the `esp32.GNUmakefile` to use xtensa toolchain if an esp32 outmodule is detected. With this change I can now successfully build the target `sled` (and libsled.a still working too) with `freertos` as os and `esp32_digitalledlib` as out mod. I also moved the `libsled.a` target to main GNUmakefile because I think it might be useful for other archs too.
Also fixed/cleaned some other stuff so most your review should be implemented now :tada: The tasks/signal are still missing. But I acutally need signal in another project so chances are good I'll look into it later on :-)

xermicus commented on src/os/os_freertos.c:
```diff
@@ -0,0 +1,121 @@
+	free(m);
+}
+
+int oscore_ncpus(void) {
```
Need to do later

xermicus commented on src/os/os_freertos.c:
```diff
@@ -0,0 +1,121 @@
+	return 1;
+}
+
+void oscore_event_signal(oscore_event ev) {
```
Will do later

vifino commented on Makefiles/esp32.GNUmakefile:
```diff
@@ -0,0 +1,18 @@
+CC := ${PLATFORMIO_HOME}/packages/toolchain-xtensa32/bin/xtensa-esp32-elf-gcc
+AR := ${PLATFORMIO_HOME}/packages/toolchain-xtensa32/bin/xtensa-esp32-elf-ar
+
+LDFLAGS ?= -nostdlib \
```
`?=` is set if not set, you want `:=`.
```suggestion
LDFLAGS := -nostdlib \
```

xermicus commented:
thx for the feedback :-)
I'll look into the signals this weekend, and then rebase for the merge

vifino commented:
Sounds good!

By the way, this only builds a library. How do I get something I can flash to an esp32?
I've found a 64x8 matrix and an esp32.

xermicus commented:
Cool!
To get a flashable binary the easiest options are to use something like esp-idf or platformio. It should be doable with esp-idf but I use platformio. Maybe it is possible to add a firmware target to the esp32 Makefile but I don't know how to it...
The following works for me (use `pio run --target upload` in the last line to directly flash it): 
```Dockerfile
FROM archlinux/base

RUN pacman -Sy
RUN pacman -Syyu --noconfirm
RUN pacman -S --noconfirm git sudo binutils coreutils python2 python2-pip tar file awk fakeroot make 
RUN useradd -m build
RUN echo "build ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers

USER build
WORKDIR /home/build
RUN git clone https://aur.archlinux.org/python-click-5.1.git
RUN git clone https://aur.archlinux.org/platformio.git
RUN git clone --single-branch --branch esp32-support https://github.com/xermicus/sled.git
RUN git clone --single-branch --branch platformio-library https://github.com/xermicus/ESP32-Digital-RGB-LED-Drivers.git

RUN cd python-click-5.1 && makepkg -si --noconfirm
RUN cd platformio && makepkg -si --noconfirm

RUN pio platform install espressif32 --with-package framework-arduinoespressif32
RUN pio lib --global install /home/build/ESP32-Digital-RGB-LED-Drivers/

RUN cd sled && \
    cp Makefiles/sledconf.esp32 sledconf && \
    make libsled.a

RUN mkdir sled_firmware && cd sled_firmware && \
	pio init && \
	echo $'[env:mhetesp32devkit] \n\
platform = espressif32 \n\
board = mhetesp32devkit \n\
framework = arduino \n\
lib_deps = /home/build/ESP32-Digital-RGB-LED-Drivers \n\
build_flags = -L/home/build/sled -lsled' > platformio.ini && \
	echo $'#include <Arduino.h> \n\
#include "esp_task_wdt.h" \n\
#include "FreeRTOS.h" \n\
extern "C" { \n\
	void user_main(void* pvParameters, int core); \n\
	int signal(int sign, void * hnd) { \n\
		return 0; \n\
 	 } \n\
} \n\
void setup() { \n\
	int pin = 22; \n\
	pinMode(pin, OUTPUT); \n\
	digitalWrite(pin, LOW); \n\
	user_main(NULL, -1); \n\
} \n\
void loop(){}' > src/main.cpp && \
	pio run
```

xermicus commented:
Gave the event tasks a quick shot. The animations are running really smooth now on my matrix, gfx_mandelbrot (which uses taskpool?) is now working.
The tasks handle is used as `oscore_event`. `oscore_event_wait_until()` blocks until an it gets unblocked by another task calling `calling oscore_event_signal()` using it's taskhandle or `desirec_usec` is reached. Can you please have a look at this? I think this might be an easy way to implement it like in os_unix but maybe I'm overlooking/misunderstanding something.

vifino commented on src/os/os_freertos.c:
```diff
@@ -86,8 +87,9 @@ void oscore_mutex_free(oscore_mutex m) {
-	return N_CPUCORES;
+	esp_chip_info_t chip_info;
+	esp_chip_info(&chip_info);
+	return chip_info.cores;
```
Oh, hey, that's pretty smart!

We should probably guard this with an ESP32 specific ifdef, or is this something freertos comes with?

xermicus replied:
Yes this is provided by esp32 sdk so a guard would make sense :+1:

vifino commented:
I don't think using the task handle as the event handle is okay. I think it'll explode sooner or later, given that the task it's running on is likely the main thread and sharing one event for all things will not really work/lead to problems.

I have hardware to test this PR out on, I'll do it when I get back from work.
For now, though, I have to switch trains, talk to you later.

vifino commented:
Hey, seems like I finally returned from work.
I am running the port on *the original sled matrix*. The first I built.

Unfortunately it was running way too fast, practically like the wait has no effect.
I fixed it, now I need to figure out why the watchdog kills it sometimes.

By the way, can I have commit rights to your branch/fork?

orithena commented:
@vifino: the watchdog reset may be due to not running WiFi. If I use WiFi.forceSleepBegin in standard Arduino, I also have to add wdt_reset() in the loop to "manually" reset the watchdog timer. Maybe that's the point?

vifino commented:
Hmm. I have changed the timers to use the Semaphore API, but while debugging why it always takes the early exit I found out that it gets called with a `desired_usec` of `1`. No clue why.

xermicus commented:
> Hmm. I have changed the timers to use the Semaphore API, but while debugging why it always takes the early exit I found out that it gets called with a `desired_usec` of `1`. No clue why. For the record, it works, but is busy-waiting.

Yeah I noticed that too but couldn't find out why. Sorry I got to pass some exams this week, so until the weekend I won't be able to do anything here anyways so feel free if you want :+1:

vifino commented:
I noticed the original author can push in forks regardless, but thanks!

Education goes first, don't worry. I'll take a look, maybe @orithena too. After all, he does more with ESP32's.

vifino commented:
BTW: The commit I pushed broke threads, but it is wrong otherwise.
We need to have multiple events created from the same thread.

I need to do more debugging there, but help would be greatly appreciated. Weird stuff.

xermicus commented:
Looks better now

vifino commented:
Do timer and threads work now?

xermicus commented:
mandelbrot2 seems to work properly on my chip but I don't have a matrix atm
are there any tests for this?

vifino commented:
Tests? ;)

vifino commented on src/os/os_freertos.c:
```diff
@@ -73,12 +73,11 @@ int oscore_event_wait_until(oscore_event ev, oscore_time desired_usec) {
-		printf("EVENT INTERRUPT!\n");
-		xSemaphoreGive(ev);
-		return 1; // we got an interrupt
+	// My experience this is more consistent
```
Of course, it is busylooping.
Ideally, we would not do this to allow other FreeRTOS threads to run, but for now this is fine.

xermicus replied:
we could make it configurable via define

xermicus commented:
@vifino ping - I'd like to get this merged since we use this code on multiple matrices :-D

vifino commented:
Same here! Perhaps it is time to just merge it.
Are you aware of any remaining issues? Like the timing still being off or something.

I'll just merge it now and we'll work on the remaining issues on master.

xermicus commented:
> Same here! Perhaps it is time to just merge it.
> Are you aware of any remaining issues? Like the timing still being off or something.
> 
> I'll just merge it now and we'll work on the remaining issues on master.

:heart:

Awesome! Right now all the initial issues are gone, it works beautifully here on two different displays. If you find something off just let me know :-)
